### PR TITLE
Fix NPE in BigQueryIO.TransformingReader when it contains an unsplittable reader.

### PR DESCRIPTION
**Pull request by peihe**

### CODE REVIEW

**Comment by peihe** [comment]
R: @dhalperi

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java [hunk 1]
```diff
@@ -673,4 +680,93 @@ public SplitAtFractionResult(int numPrimaryItems, int numResidualItems) {
         numItemsToReadBeforeSplitting, fraction, options);
     return (res.numResidualItems > 0);
   }
+
+  /**
+   * Returns an equivalent unsplittable {@code BoundedSource<T>}.
+   */
+  public static <T> BoundedSource<T> toUnsplittableSource(BoundedSource<T> boundedSource) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java]: missing a bunch of overrides on the reader, for the functions that have default implementations.

Why is this necessary?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java [hunk 2]
```diff
@@ -673,4 +680,93 @@ public SplitAtFractionResult(int numPrimaryItems, int numResidualItems) {
...
+    return new UnsplittableSource<>(boundedSource);
+  }
+
+  private static class UnsplittableSource<T> extends BoundedSource<T> {
+
+    private final BoundedSource<T> boundedSource;
+
+    private  UnsplittableSource(BoundedSource<T> boundedSource) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java]: remove extra space

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java [hunk 3]
```diff
@@ -673,4 +680,93 @@ public SplitAtFractionResult(int numPrimaryItems, int numResidualItems) {
...
+      this.boundedSource = checkNotNull(boundedSource, "boundedSource");
+    }
+
+    @Override
+    public List<? extends BoundedSource<T>> splitIntoBundles(
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java]: implement the rest of the methods on BoundedSource and BoundedReader. (there are some that are not abstract, but have default implementations).

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java [hunk 4]
```diff
@@ -673,4 +680,93 @@ public SplitAtFractionResult(int numPrimaryItems, int numResidualItems) {
...
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        return boundedReader.advance();
+      }
+
+      @Override
+      public T getCurrent() throws NoSuchElementException {
+        return boundedReader.getCurrent();
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java]: add unit tests that all methods are exposed correctly and that it rejects both initial splitting and dynamic splitting.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java (hunk 1 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java]: addressed

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java (hunk 2 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java (hunk 3 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java (hunk 4 above)
Comment (peihe) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/SourceTestUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    BoundedSource<Long> baseSource = CountingSource.upTo(100);
+    BoundedSource<Long> unsplittableSource = SourceTestUtils.toUnsplittableSource(baseSource);
+    List<?> splits = unsplittableSource.splitIntoBundles(1, options);
+    assertEquals(splits.size(), 1);
+    assertEquals(splits.get(0), unsplittableSource);
+
+    BoundedReader<Long> unsplittableReader = unsplittableSource.createReader(options);
+    assertNull(unsplittableReader.splitAtFraction(0.5));
+
+    Assert.assertEquals(0, unsplittableReader.getFractionConsumed(), 1e-15);
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/SourceTestUtilsTest.java]: use static imports here, too?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/SourceTestUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    SourceTestUtils.readNItemsFromUnstartedReader(unsplittableReader, 50);
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/SourceTestUtilsTest.java]: actually verify these? Read all elements from the unsplittable source and assert that you get 100 of them? And assert that they're equal to the ones you read from the baseSource?

**Comment by dhalperi** [comment]
Minor changes, then LGTM

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/SourceTestUtilsTest.java (hunk 5 above)
Comment (peihe) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/SourceTestUtilsTest.java]: done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/SourceTestUtilsTest.java (hunk 6 above)
Comment (peihe) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/SourceTestUtilsTest.java]: done

**Comment by dhalperi** [comment]
Beautiful. Thanks, Pei!
